### PR TITLE
new seed on redirect + error page check:

### LINKED ISSUE #475: Handle seed redirects
A seed URL may redirect to a different page via a 3xx. For prefix/domain scope crawls, this may result in the page after redirect being out of scope, resulting in crawl ending early.
eg. https://nytimes.com/ redirects to https://www.nytimes.com/, so if the former is specified as a seed with scopeType prefix, the crawl will end after the first page.

To be as flexible as possible, the crawler should check if seed page URL is different from original URL, and add the new URL as a seed as well with same scoping rules.

### PR DESCRIPTION
**Pull request by ikreymer**
- if a seed page redirects (page response != seed url), then add the final url as a new seed with same scope
- add newScopeSeed() to ScopedSeed to duplicate seed with different URL, store original includes / excludes
- also add check for 'chrome-error://' URLs for the page, and ensure page is marked as failed if page.url() starts with chrome-error://
- fixes #475

### CODE REVIEW

**Comment by benoit74** [comment]
I love it, this would greatly simplify openZIM `zimit` scraper, we do the same logic in Python and was a bit of a concern for us. FYI, our concern is that our current solution means two UAs, and even when you fake the UA header some WAF achieve to detect there is two UA thanks to some "TLS black magic".

So clearly, the sooner this is merged, the better for us ^^

File: src/crawler.ts [hunk 1]
```diff
@@ -1556,22 +1556,43 @@ self.__bx_behaviors.selectMainBehavior();
         throw new Error("page response missing");
       }
 
+      const respUrl = resp.url();
+      const isChromeError = page.url().startsWith("chrome-error://");
+
+      if (respUrl !== url && depth === 0 && !isChromeError) {
```
Comment (benoit74) [inline src/crawler.ts:1562]: Very minor suggestion, I would rewrite it `if (depth === 0 && !isChromeError && respUrl !== url ) {` for clarity

**Comment by benoit74** [comment]
Would it be easy to "inform" the user in a programmatic way about the new URL? We could detect it based on the record depth, but I don't think that the depth is saved in the WARC records. We probably need to have access to this new "fixed seed" URL at some point in our `zimit` scraper to know what is the main entry point. This would also help [replayer](https://replayweb.page/) to know where he should start indeed.

**Comment by tw4l** [comment]
> Would it be easy to "inform" the user in a programmatic way about the new URL? We could detect it based on the record depth, but I don't think that the depth is saved in the WARC records. We probably need to have access to this new "fixed seed" URL at some point in our `zimit` scraper to know what is the main entry point. This would also help [replayer](https://replayweb.page/) to know where he should start indeed.

Hm, this is a good point. Currently I think one would have to dig into the logs to find this out. Perhaps we could add a new `pages/seeds.jsonl` file or similar?

**Comment by benoit74** [comment]
> Perhaps we could add a new pages/seeds.jsonl file or similar?
This would be ok for us as well (anything else than parsing the logs is probably ok indeed 😅, just choose what suit your "vision" of the product, we can adapt 😉)

**Comment by ikreymer** [comment]
> > Perhaps we could add a new pages/seeds.jsonl file or similar?
> 
> This would be ok for us as well (anything else than parsing the logs is probably ok indeed 😅, just choose what suit your "vision" of the product, we can adapt 😉)

Hmm, could perhaps add it to the `--statsFilename` log in some way, if that's easier, although there could be more than one seed and more than one seed that redirects... I think this flag is mostly for Zimit use so can add something there. What would be most useful for Zimit? To print the actual seed used post-redirect? To indicate something like `seedRedirected: true`?

**Comment by ikreymer** [comment]
One remaining question: how should we handle a seed that redirects to a page that has already been crawled? Perhaps not as common for single-seed, but could happen with multiple seed crawls, eg.

https://someseed.example.com/
https://someotherseed.example.com/ -> https://someseed.example.com/

Should we add another `isInScope()` for the response URL, similar to: https://github.com/webrecorder/browsertrix-crawler/blob/dev-1.0.0/src/util/worker.ts#L391

**Comment by benoit74** [comment]
> Hmm, could perhaps add it to the --statsFilename log in some way, if that's easier, although there could be more than one seed and more than one seed that redirects... I think this flag is mostly for Zimit use so can add something there. What would be most useful for Zimit? To print the actual seed used post-redirect? To indicate something like seedRedirected: true?

For Zimit, `--statsFilename` is clearly the easiest solution since we already watch this file. I think we probably need the actual seed used post-redirect ; could be a list of exactly what you push in the log (origUrl, newUrl, seedID), which would handle nicely the case where multiple seeds have been passed.

> One remaining question: how should we handle a seed that redirects to a page that has already been crawled? Perhaps not as common for single-seed, but could happen with multiple seed crawls

Anyway, there will be a WARC record  in the archive for the seed that redirects, isn't it? If yes, then I would just skip adding the new seed in such a case.

**Comment by ikreymer** [comment]
> > Hmm, could perhaps add it to the --statsFilename log in some way, if that's easier, although there could be more than one seed and more than one seed that redirects... I think this flag is mostly for Zimit use so can add something there. What would be most useful for Zimit? To print the actual seed used post-redirect? To indicate something like seedRedirected: true?
> 
> For Zimit, `--statsFilename` is clearly the easiest solution since we already watch this file. I think we probably need the actual seed used post-redirect ; could be a list of exactly what you push in the log (origUrl, newUrl, seedID), which would handle nicely the case where multiple seeds have been passed.
> 
> > One remaining question: how should we handle a seed that redirects to a page that has already been crawled? Perhaps not as common for single-seed, but could happen with multiple seed crawls
> 
> Anyway, there will be a WARC record in the archive for the seed that redirects, isn't it? If yes, then I would just skip adding the new seed in such a case.

Yes, there will be, so when parsing the WARC you can just follow the redirect chain for the seed until you find the non-redirect. Maybe a bit more work but should be straightforward since you're verifying that the WARC has the seed already. That might be the simplest option, since the statsFilename output is the stats on the current state of the crawl, not about the seed.

**Comment by ikreymer** [comment]
> One remaining question: how should we handle a seed that redirects to a page that has already been crawled? Perhaps not as common for single-seed, but could happen with multiple seed crawls, eg.
> 
> https://someseed.example.com/ https://someotherseed.example.com/ -> https://someseed.example.com/
> 
> Should we add another `isInScope()` for the response URL, similar to: https://github.com/webrecorder/browsertrix-crawler/blob/dev-1.0.0/src/util/worker.ts#L391
> 
> If the page has already been crawled, we just skip the page? That would result in behaviors / screenshots being skipped, so may be confusing. Or, maybe we just commit to full capture of any page that is specified as a seed.
> 
> Thoughts?

Decided to keep current behavior, as user supplied seeds are guaranteed to be loaded and processed fully. Can perhaps as a separate seed dedup option later if necessary.

File: src/crawler.ts (hunk 1 above)
Comment (ikreymer) [reply src/crawler.ts:1562]:
```suggestion
      if (depth === 0 && !isChromeError && respUrl !== url) {
```

**Comment by benoit74** [comment]
> Yes, there will be, so when parsing the WARC you can just follow the redirect chain for the seed until you find the non-redirect. Maybe a bit more work but should be straightforward since you're verifying that the WARC has the seed already. That might be the simplest option, since the statsFilename output is the stats on the current state of the crawl, not about the seed.

This is enough indeed, no need for more I think (this PR is already a great step forward for us)

**Pull request merged**